{% extends "base.html" %}
{% load media %}
{% block title %}About NPACT.{% endblock %}

{%block content %}
<div class="contentPadding">
<!--<h2 id="secHeader" title="NPACT - About"><span>--><!--header image here--><!--</span>NPACT About</h2> -->
<img src="{% url start %}devassets/images/header1.jpg" title="NPACT - About" alt="NPACT - About" style="margin:0 0 5px 8px; width:98%" />
<p></p><h4>About</h4>
<p>NPACT is a computational and graphical representation tool for gene identification and sequence annotation. NPACT identifies sequence segments of any length with <b>statistically-significant 3-base compositional periodicities</b> and <b>associated with ORF structures</b> (<a href="">Oden and Brocchieri 2014</a>, submitted). It uses a new quantitative approach that extends the principle of visual frame analysis (<a href="http://www.sciencedirect.com/science/article/pii/0378111984901161">Bibb et al. 1984</a>).</p>

<p>The NPACT server produces a multi-page graphical output representing, along sequence positions, frame-specific GC content (or optionally content of any nucleotide combination) for frame analysis (<a href="http://www.bioinformatika.pbf.hr/Articles/FramePlot.pdf">Ishikawa and Hotta 1999</a>; <a href="http://www.ncbi.nlm.nih.gov/pubmed/15919911">Brocchieri et al. 2005</a>), published gene annotations (when available), hits, and newly identified ORFs with 3-base compositional periodicities.</p>

<p>NPACT can be used for characterizing and editing pre-existing annotations, or for de novo gene finding in prokaryotic sequences. In de novo gene-finding mode, NPACT will be most useful for identifying coding regions in sequences that are short or of very variable composition, and in sequences of high GC content (>60%). NPACT is also useful for characterizing gene predictions from other methods, such as prokaryotic genome annotations, or generated by any other predictor, using a Genbank-style annotation and sequence input file. The multiple information provided by the NPACT graphical output and text files will be useful to corroborate hypothetical gene predictions and to identify a manageable number of possible problem areas for further analysis (e.g., conservation analysis).</p>

<h4>Getting Started</h4>
<p>To start the analysis first upload a nucleotide sequence file in the Genbank format, using any of the methods provided. Directly start the analysis with default parameters clicking "Get Analysis!", or change options with "Configure Analysis". The program accepts as input a Genbank-like sequence text file that <b>must</b> include a <b>multi-line header</b> minimally including a <b>LOCUS</b> first line containing sequence length information in the Genbank format, followed by a line starting with <b>DEFINITION</b>, which may contain species name, and finally a line starting with <b>ORIGIN</b>, followed by lines containing sequence data. All <b>alphabetic</b> characters are interpreted as sequence characters. Any alphabetic character other than A/a, C/c, G/g, T/t, U/u is treated as unknown. Non-alphabetic characters are ignored. Optional annotated coding sequences are read from the header's " CDS " fields, following the Genbank format. Gene names are read from the variables "gene" (first choice) or "locus_tag" (second choice).</p>

<h4>N-profile and prediction graphical output</h4>
<p>NPACT produces a multi-page graphical ouput in Postscript or PDF format, containing the following information:</p>
<p> <b>Newly Identified ORFs</b>. These are ORFs newly identified by the NPACT procedure (not included in the submitted CDS annotations) or potentially significantly modifying start of translation of an annotated CDS. ORFS are represented as pointing in the 5'-to-3' orientation (see Coloring Scheme below coloring-key). Newly identified ORFs are in bold-face, whereas those matching coding regions corresponding to published genes but with modified starting position are displayed in lighter color. The names assigned to the newly identified ORFs are in the form:</p><p>[H/G]-####[-/*][A/a/G/g/T/t/C/c/W/w]</p><p>For example: H-2335*A, G-123-t, where:</p>
<list>
<p>	<b>[H/G]</b> identifies the type of hit:</p>
<list>
<p>		<b>H</b>: H-type hit, identified using a scoring scheme based on compositional expectations of coding regions.</p>
<p>		<b>G</b>: G-type hit, identified by any non-random association of base type with codon position.</p>
</list>
<p>	<b>####</b>: A unique identification number followed by:</p>
<p>	<b>[-/*]</b> identifies the E-value:</p>
<list>
<p>	<b>-</b>: E-value > 0.01 (weaker compositional periodicity).</p>
<p>	<b>*</b>: E-value >= 0.01 (stronger compositional periodicity).</p>
</list>
<p>	<b>[A/a/G/g/T/t/C/c/W/w]</b> indicates the potential start-of-translation codon closest to the 5'-end of the hit. Upper case indicates that the predicted codon is within 15 codons of the start of the hit (and up to three codons within the hit), lower case indicates that it is more distant. The potential start-of-translation codons considered are, in the order:</p>
<list>
<p>		ATG (A/a) or GTG (G/g)</p>
<p>		TTG (T/t)</p>
<p>		CTG (C/c)</p>
<p>		ATT (W/w)</p>
</list>
<p><b>Input file CDS</b>. Coding regions reported in the annotation of the Genbank input file, from the CDS field (see Coloring Scheme below).</p>
<p><b>Hits</b>. Graphical representation of sequence segments with compositional periodicity identified as colored segments on two tracks, corresponding to direct strand (upper track) and complementary strand (lower track). See COroring Scheme below for color assignment.</p>
<p><b>N-Profiles</b>. These curves represent by default GC content profiles (S-Profiles) within a moving window of 201 nucleotides. The user can choose to display any other nucleotide combination (e.g., purines: A+G) with the understanding that any combination other than C+G or A+T will show nucleotide frequencies on the direct strand and the frequency of the complementary bases on the complementary strand. Each N-profile represents percent frequencies within a subsequence composed of every third nucleotide of the input sequence, starting, respectively, from position 1 (red), 2 (green) or 3 (blue). A light-gray curve indicates the average composition within the corresponding interval. Alternative nucleotide-type combinations (e.g., purines) can optionally be displayed from the Configure Analysis page.</p>
</list>

<h4>Coloring Scheme</h4>
<p>The coloring scheme used throughout the graphical output follows the following rules. The nucleotide sub-sequences starting, respectively, from position 1, 2 or 3 of the sequence - composed of positions (<it>i</it> mod 3) where <it>i</it> = 1, 2 or 3 - are colored in red (R), green (G), and blue (B), respectively. Annotated genes, newly identified ORFs, and hits, are colored as the sub-sequence in phase with their third codon positions. An alternative coloring scheme is available.</p>
</div>

{%endblock content %}

