{% extends "base.html" %}
{% load media %}
{% block title %}About N-PACT.{% endblock %}

{%block content %}
<div class="contentPadding">
<h2 id="secHeader" title="N-PACT - About"><span><!--header image here--></span>N-PACT About</h2>
<p>You reached a version under development of N-PACT, a program designed to identify protein-coding regions by frame analysis (Bibb et al, 1984), on a genomic scale. N-PACT produces a multi-page graphical representation of frame-specific compositional properties of the sequence (frame analysis), together with representation of coding regions annotated in the input sequence file. It also performs quantitative frame analysis on the sequence, identifying sequence regions of significant compositional contrasts and corresponding gene predictions (<a href="">Oden and Brocchieri 2012</a>). </p>

<p>The N-PACT server produces a (multi-page) PostScript file for frame analysis (<a href="http://www.sciencedirect.com/science/article/pii/0378111984901161">Bibb et al. 1984</a>) representing profiles of GC content in three frames (<a href="http://www.bioinformatika.pbf.hr/Articles/FramePlot.pdf">Ishikawa and Hotta 1999</a>; <a href="http://www.ncbi.nlm.nih.gov/pubmed/15919911">Brocchieri et al. 2005</a>), published gene annotation when available, high-scoring segments and new gene predictions.</p>

<h4>Getting Started</h4>
<p>To start the analysis upload a nucleotide sequence file in the GenBank format using any of methods provided. Directly start the analysis with default parametrs with "Get Analysis!" or "Configure Analysis" before launching it. The program accepts as input a GenBank-formatted sequence text file that minimally must include one line starting with the word ORIGIN, followed by lines with the nucleotide sequence, in lower or upper case. All alphabetic characters following the ORIGIN line are interpreted as sequence characters (any alphabetic character other than A/a, C/c, G/g, T/t, U/u, is interpreted as unknown), and all non-alphabetic characters are ignored. All annotated coding sequences are read from the file up to the line ORIGIN according to the GenBank format, from " CDS " lines. Gene names are assigned reading them from the variables "gene" (first choice) or "locus_tag" (second choice).</p>

<h4>N-profile and prediction PostScript output</h4>
<p>The multi-page PostScript ouput file produced by N-PACT contains the following information:</p>
<p> Newly Annotated. These are coding regions predicted by the N-PACT procedure, represented by arrows pointing in the 5'-to-3' direction. Newly predicted regions are in bold-face whereas regions corresponding to published genes but modified in their starting position are displayed in lighter color. The names of the coding regions are in the form: [H/G]-number[-/*]-[A/a/G/g/T/t/C/c/W/w/], with the following meaning:</p>
<list>
<p>	H: H-type hits, identified using a scoring scheme.</p>
<p>	G: G-type hits, identified by non-random association of base type with codon position.</p>
<p>	number: A unique identification number followd by:</p>
<p>	- if significance does not reach global threshold (less strong contrast).</p>
<p>	* if ignificance reaches global threshold (stronger contrast).</p>
<p>	The following upper or lower case character indicates a predicted start-of-translation codon. Upper case indicates that the predicted codon is within 15 positions of the start of the hit, lower case indicates that it is more distant. The potential start-of-translation codons looked for are, in the order:</p>
<list>
<p>		ATG (A/a) or GTG (G/g)</p>
<p>		TTG (T/t)</p>
<p>		CTG (C/c)</p>
<p>		AAT (W/w)</p>
</list>
<p>Annotation. These are coding regions reported in the annotation of the GenBank input file, from the CDS field (see above).</p>
<p>Hits. Graphical representation of high scoring segments identified as colored segments on to tracks, corresponding to direct strand (upper track) and complementary strand (lower track).</p>
<p>Profiles. These curves represent by default GC content within a moving window of 201 nucleotides. Usage of other nucleotide type combinations can be chosen by the user. Each curve represents percent frequencies within a subsequence composed of every third nucleotide of the genome, starting, respectively, from position 1 (red), 2 (green) or 3 (blue). A light-gray curve indicates the overall composition of the genome within corresponding intervals.</p>
</list>

<h4>Coloring scheme</h4>
<p>The coloring scheme used throughout the output follows the following rules. The nucleotide sequence is partitioned into three subsequences composed of every third nucleotide, starting, respectively, from position 1, 2 and 3 of the sequence. These subsequences - composed of positions (i mod 3), where i = 1, 2 or 3 - are assigned different colors, e.g., red (R), grenn (G), and blue (B). Annotated genes, newly predicted genes, and high-scoring segments, are assigned the same color of the subsequence to which their third codon positions belong.</p>
</div>

{%endblock content %}

