{% extends "base.html" %}
{% load media %}
{% block title %}About N-P.A.C.T.{% endblock %}

{%block content %}
<div class="contentPadding">
<p>You reached a version under development of N-PACT. N-PACT provides: (1) a genome-scale visual representation of sequence compositional properties for the identification of protein-coding sequences in relation to published annotation and, by default, position of high-scoring-segments (HSSs) of significant compositional contrast and corresponding gene predictions using the method of Oden and Brocchieri (<a href="">in preparation</a>). </p>

<p>The N-PACT server produces a (multi-page) PostScript file for frame analysis (<a href="http://www.sciencedirect.com/science/article/pii/0378111984901161">Bibb et al. 1984</a>) representing profiles of GC content in three frames (<a href="http://www.bioinformatika.pbf.hr/Articles/FramePlot.pdf">Ishikawa and Hotta 1999</a>; <a href="http://www.ncbi.nlm.nih.gov/pubmed/15919911">Brocchieri et al. 2005</a>), published gene annotation when available, high-scoring segments and new gene predictions.</p>

<h4>Getting Started</h4>
<p>To start the analysis click "Run N-Profile Analysis" at the upper-right corner, input sequence file, click "Start Analysis!", choose parameters, and finally click "Get Analysis!". The program accepts as input a GenBank-formatted sequence text file, which has to minimally include one line starting with the word ORIGIN, followed by lines including the nucleotide sequence in lower or upper case. All alphabetic characters are interpreted as sequence, and all non-alphabetic characters are ignored. All annotated coding sequences are read from the file up to the line ORIGIN according to the GenBank format, from " CDS " lines. Gene names are assigned reading them from the variables "gene" (first choice) or "locus_tag" (second choice).</p>
</div>

{%endblock content %}

