{% extends "base.html" %}
{% load media %}
{% block title %}About N-P.A.C.T.{% endblock %}

{%block content %}
<div class="contentPadding">
<p>You reached a version under development of N-PACT, a program designed to identify protein-coding regions by frame analysis (Bibb et al, 1984), on a genomic scale. N-PACT produces a multi-page graphical representation of frame-specific compositional properties of the sequence (frame analysis), together with representation of coding regions annotated in the input sequence file. It also performs quantitative frame analysis on the sequence, identifying sequence regions of significant compositional contrasts and corresponding gene predictions (<a href="">Oden and Brocchieri 2012</a>). </p>

<p>The N-PACT server produces a (multi-page) PostScript file for frame analysis (<a href="http://www.sciencedirect.com/science/article/pii/0378111984901161">Bibb et al. 1984</a>) representing profiles of GC content in three frames (<a href="http://www.bioinformatika.pbf.hr/Articles/FramePlot.pdf">Ishikawa and Hotta 1999</a>; <a href="http://www.ncbi.nlm.nih.gov/pubmed/15919911">Brocchieri et al. 2005</a>), published gene annotation when available, high-scoring segments and new gene predictions.</p>

<h4>Getting Started</h4>
<p>To start the analysis upload a nucleotide sequence file in the GenBank format using any of methods provided. Directly start the analysis with default parametrs with "Get Analysis!" or "Configure Analysis" before launching it. The program accepts as input a GenBank-formatted sequence text file, which, minimally must include one line starting with the word ORIGIN, followed by lines with the nucleotide sequence, in lower or upper case. All alphabetic characters following the ORIGIN line are interpreted as sequence characters (any alphabetic character other than A/a, C/c, G/g, T/t, U/u, is interpreted as unknown), and all non-alphabetic characters are ignored. All annotated coding sequences are read from the file up to the line ORIGIN according to the GenBank format, from " CDS " lines. Gene names are assigned reading them from the variables "gene" (first choice) or "locus_tag" (second choice).</p>
</div>

{%endblock content %}

